swift/ql/lib/codeql/swift/elements/AstNode.qll

```diff
@@ -21,10 +23,26 @@ private module Cached {
   AbstractFunctionDecl getEnclosingFunction(AstNode ast) {
     result = getEnclosingFunctionStep*(getEnclosingDecl(ast))
   }
+
+  private Element getEnclosingClosureStep(Element e) {
+    not e instanceof Callable and
+    result = getImmediateParent(e)
+  }
+
+  cached
+  AbstractClosureExpr getEnclosingClosure(AstNode ast) {
+    result = getEnclosingClosureStep*(getImmediateParent(ast))
+  }
 }
 
 class AstNode extends Generated::AstNode {
   final AbstractFunctionDecl getEnclosingFunction() { result = Cached::getEnclosingFunction(this) }
 
   final Decl getEnclosingDecl() { result = Cached::getEnclosingDecl(this) }
+
+  final Callable getEnclosingCallable() {
+    if exists(Cached::getEnclosingClosure(this))
+    then result = Cached::getEnclosingClosure(this)
+    else result = Cached::getEnclosingFunction(this)
+  }
```

Comment:
So if I have something like:
```swift
func foo() {
  var f = {(n : Int) in n + 1 }
}
```
the result of calling `getEnclosingCallable` on the `n + 1` AstNode will be the Callable that represents `{(n : Int) in n + 1 }`, and the result of calling `getEnclosingFunction` will be `foo`?

(If my reading is correct, this should probably be converted into some QLDoc.)

I can totally see the value of having this distinction. It would be good to check if this is the distinction other languages with first-class functions also make.

Reply:
Correct.

I don't know if it makes sense to have getEnclosingFunction publicly at this point, but its implementation is optimized, so I thought it would be worth keeping separately anyway.

Also many functions and operators take autoclosures, so in many cases getting that instead of a named function would be surprising.